core/bentley/src/StatusCategory.ts

```diff
@@ -0,0 +1,355 @@
+    case ChangeSetStatus.MergeSchemaChangesOnOpen: return new Directive();
+    case ChangeSetStatus.ReverseOrReinstateSchemaChanges: return new Directive();
+    case ChangeSetStatus.ProcessSchemaChangesOnOpen: return new Directive();
+    case ChangeSetStatus.CannotMergeIntoReadonly: return new ValidationError();
+    case ChangeSetStatus.CannotMergeIntoMaster: return new ValidationError();
+    case ChangeSetStatus.CannotMergeIntoReversed: return new ValidationError();
+
+    case RepositoryStatus.Success: return new Success();
+    case RepositoryStatus.ServerUnavailable: return new NetworkError();
+    case RepositoryStatus.LockAlreadyHeld: return new Conflict();
+    case RepositoryStatus.SyncError: return new NetworkError();
+    case RepositoryStatus.InvalidResponse: return new NetworkError();
+    case RepositoryStatus.PendingTransactions: return new StateViolation();
+    case RepositoryStatus.LockUsed: return new StateViolation();
+    case RepositoryStatus.CannotCreateChangeSet: return new InternalError();
+    case RepositoryStatus.InvalidRequest: return new NetworkError();
+    case RepositoryStatus.ChangeSetRequired: return new StateViolation();
+    case RepositoryStatus.CodeUnavailable: return new Conflict();
+    case RepositoryStatus.CodeNotReserved: return new StateViolation();
+    case RepositoryStatus.CodeUsed: return new StateViolation();
+    case RepositoryStatus.LockNotHeld: return new Forbidden();
+    case RepositoryStatus.RepositoryIsLocked: return new Locked();
+    case RepositoryStatus.ChannelConstraintViolation: return new ConstraintViolation();
+
+    case HttpStatus.Success: return new Success();
+
+    case IModelHubStatus.Success: return new Success();
+    case IModelHubStatus.Unknown: return new UnknownError();
+    case IModelHubStatus.MissingRequiredProperties: return new ValidationError();
+    case IModelHubStatus.InvalidPropertiesValues: return new ValidationError();
+    case IModelHubStatus.UserDoesNotHavePermission: return new PermissionsViolation();
+    case IModelHubStatus.UserDoesNotHaveAccess: return new PermissionsViolation();
+    case IModelHubStatus.InvalidBriefcase: return new ValidationError();
+    case IModelHubStatus.BriefcaseDoesNotExist: return new NotFound();
+    case IModelHubStatus.BriefcaseDoesNotBelongToUser: return new PermissionsViolation();
+    case IModelHubStatus.AnotherUserPushing: return new StateViolation();
+    case IModelHubStatus.ChangeSetAlreadyExists: return new Conflict();
+    case IModelHubStatus.ChangeSetDoesNotExist: return new NotFound();
+    case IModelHubStatus.FileIsNotUploaded: return new StateViolation();
+    case IModelHubStatus.iModelIsNotInitialized: return new StateViolation();
+    case IModelHubStatus.ChangeSetPointsToBadSeed: return new InvalidData();
+    case IModelHubStatus.OperationFailed: return new OperationFailed();
+    case IModelHubStatus.PullIsRequired: return new StateViolation();
```

Comment:
I am a bit concerned about how we're using `iModelHubStatus` here...  It's not that I disagree with these mappings per se, it's that I think there's a whole class of these errors that should probably _never_ be returned to the client from a backend.

For example, with something like `PullIsRequired`, I would expect a backend to always handle that response from the hub correctly, and if it doesn't, that should be treated as a server error.

I guess my real question is, are we confident that we do enough error checking internally before returning these to the client?  Since 5xx errors are likely to become the only ones that really get our attention, I think we should almost err on the side of 500s whenever we're not sure whether the client or server is at fault...

Reply:
You have a good point. Maybe 500 is not a perfect code, but it will get attention.

Thanks for the deep review. It really helps to get another pair of eyes on this stuff.